brave/brave-core
# Register wallet data files component only if user has created wallets

Resolves https://github.com/brave/brave-browser/issues/33757
Resolves https://github.com/brave/brave-browser/issues/30940

This PR does following things:
  - For normal users without wallet: wallet data files won't be registered.
  - For first time wallet users, wallet data files component will be registered
      when Create/ResotreWallet and ImportFromExternalWallet APIs are called, the
      finish callback of these APIs won't be run until wallet data files component
      is ready and content is parsed into BlockchainRegistry.
  - For existing wallet users, the behavior is unchanged, wallet data files
      component will continue to be registered and parsed upon startup.
  - Remove json sanitization on JSON files from brave wallet data
      files component served via component updater which is considered trusted data.

Note that if the installation of component or parsing failed, we won't retry and will unblock the APIs without error in this case.



## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Screenshots:
Android - Create Wallet|Android - Restore Wallet
--|--
https://github.com/brave/brave-core/assets/24739341/3fe5936c-8f51-450f-96e8-9a92d2de4af5|https://github.com/brave/brave-core/assets/24739341/76d2a094-f5e4-4710-a52f-47b3fd12bc80

### Create New Wallet

https://github.com/brave/brave-core/assets/40611140/585bd4b6-8c6d-4b9e-9850-3941c54152d3

### Import From Existing Wallet

https://github.com/brave/brave-core/assets/40611140/b5a032e2-4dad-405a-9e5d-32ed63543f87

### Restore From Seed Phrase

https://github.com/brave/brave-core/assets/40611140/145d7b1e-2fe6-4c37-b0ff-b6b01eb7ae79

## Test Plan:
1. Start with a new user data dir folder (clean state), `npm start -- --user_data_dir_name=PR20741` or run the executable with `--user-data-dir=some_path`. (Should see a welcome page when started)
2. Go to brave://components, Brave Wallet data files should not present.
3. Go to brave://wallet, create a new wallet
4. Visit brave://wallet/crypto/portfolio/add-asset#available-assets, should see tokens, and Brave Wallet data files should present in brave://components
5. Start with another new user data dir folder
6. Go to brave://components, Brave Wallet data files should not present.
7. Go to brave://wallet, import from an existing wallet to restore from seed phrase.
8. Visit brave://wallet/crypto/portfolio/add-asset#available-assets, should see tokens, and Brave Wallet data files should present in brave://components

Also, do the similar steps for testing `import from existing wallet` like above video shows.
